Build/python3 migration

BonfaceKilz opened the pull request:
#### Description

This PR:
- Migrates python2 to python3 using the tool `2to3.8`
- Fixes failing tests(Mechanical Rob and unittests) along the way
- Removes unused imports in some cases
- Applies pep8 in some cases(like replacing tabs with 4 spaces)
- Replaces python2-htmlgen to python3-htmlgen

#### How should this be tested?

- Install python3-genenetwork from guix(see below). This has all the python3 packages you need.
- Confirm that all unittests pass by running:

```bash
# Replace paths where appropriate
env GN2_PROFILE=/home/bonface/opt/python3-genenetwork2 TMPDIR=/home/bonface/tmp SERVER_PORT=5004 GENENETWORK_FILES=/home/bonface/gnu/data/gn2_data/ /home/bonface/projects/genenetwork2/bin/genenetwork2 /home/bonface/projects/genenetwork2/etc/default_settings.py -c -m unittest discover -v
```

- Confirm that Mechanical Rob works:

```bash
# Replace paths and host where appropriate
time env GN2_PROFILE=/home/bonface/opt/genenetwork2 TMPDIR=/home/bonface/tmp SERVER_PORT=5004 GENENETWORK_FILES=/home/bonface/gnu/data/gn2_data/ ./bin/genenetwork2 ./etc/default_settings.py -c /home/bonface/projects/genenetwork2/test/requests/test-website.py -a http://localhost:5004
```

- Run the site(the proxy has to be up) and check that things work as expected:
```
cd /home/bonface/projects/genenetwork2 && env SERVER_PORT=5004 TMPDIR=/home/bonface/tmp WEBSERVER_MODE=DEBUG LOG_LEVEL=DEBUG GENENETWORK_FILES=/home/bonface/gnu/data/gn2_data/
GN2_PROFILE=/home/bonface/opt/python3-genenetwork2 /home/bonface/projects/genenetwork2/scripts/run_debug.sh
```

#### Any background context you want to provide?

- *gunicorn* was not working for me(still don't know why), so I used a version provided by archlinux.
- Right now, things just work. I don't know if they work correctly though. There will be some bugs/ errors that me and @zach will iteratively have to work on. But we have something that works!
- I caught some bugs through the few existing unittests that evaded 2to3.8 like https://github.com/genenetwork/genenetwork2/pull/422/commits/7717f0d3d4802d115a8b5ace4d379864ba5b1188. I'm sure there are some subtle bugs lurking around, so we'd have to be wary of those.

#### What are the relevant pivotal tracker stories?

- fixes genenetwork/genenetwork2#424
- fixes genenetwork/genenetwork2#398

#### Screenshots (if appropriate)

N/A

#### Questions

- I've had to do some hacks to make the auth stuff work:
  - https://github.com/genenetwork/genenetwork2/pull/422/commits/70dbeeb5832711ed5271434e482c18bc7ea095b8
  - https://github.com/genenetwork/genenetwork2/pull/422/commits/b46faf479e129801f4bec2457444b5341e8aeabf
  - https://github.com/genenetwork/genenetwork2/pull/422/commits/46443ec8d2cdfd7c60358a889d90a90e4f7daaf4
  - https://github.com/genenetwork/genenetwork2/pull/422/commits/7f623509a4afded47eb8580451c146487d2ef662

- Also, we have to check if the utf-8 bugs still exist. Python3 handles utf-8 strings a bit differently. 

Could @zsloan have a look at the above and give me some feedback?

#### How to install python3-genenetwork from guix

Recommended: install python3-genenetwork in its own profile. In this case, we have guix-bioinformatics and guix-past set up locally. You could always add this to it's own *channel.scm*` file.

```bash
env GUIX_PACKAGE_PATH="/home/bonface/projects/guix-bioinformatics:/home/bonface/projects/guix-past/modules" guix install python3-genenetwork2 -p ~/opt/python3-genenetwork2

```

BonfaceKilz committed:
Wrap print statements in parentheses

BonfaceKilz committed:
Apply pep8

* scripts/maintenance/readProbeSetSE_v7.py: Apply pep8 to file to replace tabs
with spaces and use correct indentation.

BonfaceKilz committed:
Replace DOS style line endings with UNIX style ones

* scripts/maintenance/readProbeSetSE_v7.py: Run *dos2unix* against file

BonfaceKilz committed:
Fix malformed docstring

* wqflask/utility/Plot.py: Update docstring.

BonfaceKilz committed:
Replace "<>" with "!="

* wqflask/utility/svg.py: Use "!=". The operator, "<>" is removed in Python3

BonfaceKilz committed:
Replace "raw_input" with "input"

* scripts/maintenance/readProbeSetMean_v7.py: Use "input". See
https://docs.python.org/3/whatsnew/3.0.html#builtins

BonfaceKilz committed:
Apply pep8

* wqflask/utility/svg.py: Apply pep8 to fix indentation error when running
`2to3-3.8 -w .`:
````
RefactoringTool: Can't parse ./wqflask/utility/svg.py: IndentationError: unindent does not match any outer indentation level (<tokenize>, line 403)
````

BonfaceKilz committed:
Wrap `map()` in a `list` call

Run `2to3-3.8 -f map -w .`

See: <https://docs.python.org/2/library/2to3.html#2to3fixer-map>

BonfaceKilz committed:
Remove erroneous `if .. else` branch

* wqflask/utility/svg.py [roct, ellipse, SVGelement]: Raise only a single value error
if either height or width is not defined. Fixes parsing error when running
`2to3-3.8 -f apply -w .`

BonfaceKilz committed:
Fix dictionary iteration methods

Run `2to3-3.8 -f dict -w .`

See: <https://docs.python.org/2/library/2to3.html#2to3fixer-dict> and
<https://stackoverflow.com/questions/17695456/why-does-python-3-need-dict-items-to-be-wrapped-with-list>

BonfaceKilz committed:
Wrap `filter()` usage in a `list` call

Run `2to3-3.8 -f filter -w .`

See <https://docs.python.org/2/library/2to3.html#2to3fixer-filter>

BonfaceKilz committed:
Change `dict.has_key(key)` to `key in dict`

Run `2to3-3.8 -f has_key -w .`

See: <https://docs.python.org/2/library/2to3.html#2to3fixer-has_key>

BonfaceKilz committed:
Replace `izip` with python's built-in equivalent

Run `2to3-3.8 -f itertools -w . && 2to3-3.8 -f itertools_imports -w .`

See: <https://docs.python.org/2/library/2to3.html#2to3fixer-itertools_imports>
and <https://docs.python.org/2/library/2to3.html#2to3fixer-itertools>

BonfaceKilz committed:
Convert the old not-equal syntax, <>, to !=

Run `2to3-3.8 -f ne -w .`

See: <https://docs.python.org/2/library/2to3.html#2to3fixer-ne>

BonfaceKilz committed:
Convert the use of iterator’s next() methods to the next() function

Run `2to3-3.8 -f next -w .`

See: <https://docs.python.org/2/library/2to3.html#2to3fixer-next>

BonfaceKilz committed:
Wrap `raise` statements in parenthesis

Run `2to3-3.8 -f raise -w .`

See: <https://docs.python.org/2/library/2to3.html#2to3fixer-raise>

BonfaceKilz committed:
Convert `raw_input` to `input`

Run `2to3-3.8 -f raw_input -w .`

See: <https://docs.python.org/2/library/2to3.html#2to3fixer-raw_input>

BonfaceKilz committed:
Handle the move of reduce() to functools.reduce()

Run `2to3-3.8 -f reduce -w .`

See: <https://docs.python.org/2/library/2to3.html#2to3fixer-reduce>

BonfaceKilz committed:
Replace backtick repr with the repr() function

* wqflask/base/trait.py: See
<https://docs.python.org/2/library/2to3.html#2to3fixer-repr>

BonfaceKilz committed:
Handle the rename of urllib and urllib2 to the urllib package

See: <https://docs.python.org/2/library/2to3.html#2to3fixer-urllib>

BonfaceKilz committed:
Rename xrange() to range() and wrap existing range() calls with list

See: <https://docs.python.org/2/library/2to3.html#2to3fixer-xrange>

BonfaceKilz committed:
Make Python more idiomatic

Run `2to3-3.8 -f idioms -w .`

See: <https://docs.python.org/2/library/2to3.html#2to3fixer-idioms>

BonfaceKilz committed:
Remove extra whitespace(or add it) from comma separated items

See: <https://docs.python.org/2/library/2to3.html#2to3fixer-urllib>

BonfaceKilz committed:
Add backward support for urllib

* wqflask/base/trait.py: Remove unused imports
* wqflask/db/call.py (Bench): Add urllib2 fallback for python2

BonfaceKilz committed:
Disable python2 QTLReaper

* wqflask/base/data_set.py: Remove usage of "reaper" to load datasets

BonfaceKilz committed:
Replace all references to python2.7 with python3.8

Run:

```
find . -type f -print0 | xargs -0 sed -i \
"s|lib/python2.7/site-packages|lib/python3.8/site-packages|g"
```

BonfaceKilz committed:
Handle module renames in the standard library

Run:
```
2to3-3.8 -f imports -w . && \
2to3-3.8 -f imports2 -w .
```

See: <https://docs.python.org/2/library/2to3.html#2to3fixer-imports> and
<https://docs.python.org/2/library/2to3.html#2to3fixer-imports2>

BonfaceKilz commented:
Hit a problem when building python-rpy2(used in guix [here](http://git.genenetwork.org/guix-bioinformatics/guix-bioinformatics/commit/fd141c17686f54d94dd68391298347de6007cf81)); but it's now fixed upstream: https://debbugs.gnu.org/cgi/bugreport.cgi?bug=42956

BonfaceKilz committed:
Remove unused imports

BonfaceKilz committed:
Use Python3 string methods

* wqflask/utility/webqtlUtil.py (genRandStr): Replace "string.letters" with
"string.ascii_letters".

BonfaceKilz committed:
Merge branch 'testing' into build/python3-migration

BonfaceKilz committed:
Remove unused imports

BonfaceKilz committed:
Update corestats import

* wqflask/utility/Plot.py: import corestats from the utility module.

BonfaceKilz committed:
Replace python2-htmlgen with python-htmlgen

* wqflask/wqflask/marker_regression/display_mapping_results.py: Use
python-htmlgen

BonfaceKilz committed:
Fix failing tests that use python2-htmlgen

* wqflask/tests/wqflask/marker_regression/test_display_mapping_results.py: Fix
tests.

BonfaceKilz committed:
Replace calls to "basestring with "str""

See: <https://docs.python.org/2/library/2to3.html#2to3fixer-basestring>

BonfaceKilz committed:
Rename "unicode" to "str"

See: <https://docs.python.org/2/library/2to3.html#2to3fixer-unicode>

BonfaceKilz committed:
Replace "__cmp__" with "__eq__" magic method

* wqflask/base/GeneralObject.py: "__cmp__" is no longer used in python3. See:
<https://stackoverflow.com/questions/8276983/why-cant-i-use-the-method-cmp-in-python-3-as-for-python-2>

BonfaceKilz committed:
Remove "from __future__ import new_feature" statements

See: <https://docs.python.org/2/library/2to3.html#2to3fixer-future>

BonfaceKilz committed:
Remove debug statement

* wqflask/wqflask/views.py(search_page): "debugf" raises an error since the
first arg is expected to be an int

BonfaceKilz committed:
Replace tabs with spaces

* scripts/maintenance/readProbeSetMean_v7.py: Run python-autopep8 on this file.

BonfaceKilz committed:
Replace "string.split" & "string.join" with python's inbuilt methods

BonfaceKilz committed:
Remove unused imports

BonfaceKilz committed:
Add global method to convert binary string to plain string

* wqflask/utility/db_tools.py: escape_string returns a binary string which
introduces a bug when composing sql query string. The escaped strings have to be
converted to plain text.

BonfaceKilz committed:
Remove redundant methods which are defined elsewhere

* wqflask/base/data_set.py: Delete "create_in_clause", and "mescape" which are
defined in "utility.db_tools"

BonfaceKilz committed:
Replace mysqldb's "escape" with the one from utils.db_tools

Mysqldb's escape returns a binary string. Use utils.db_tools which adds an extra
step of converting the binary string to a string.

BonfaceKilz committed:
Remove comment block

BonfaceKilz committed:
Apply pep-8 formatting

BonfaceKilz committed:
Fix failing tests

BonfaceKilz committed:
Replace string arguments to "hmac.new" with bytearray

See: <https://stackoverflow.com/questions/31848293/python3-and-hmac-how-to-handle-string-not-being-binary>

BonfaceKilz committed:
Import "reduce" without a try block

* wqflask/wqflask/correlation_matrix/show_corr_matrix.py: Import 'reduce'

BonfaceKilz committed:
Convert sibling imports to relative imports

See: <https://docs.python.org/2/library/2to3.html#2to3fixer-import>

BonfaceKilz committed:
Replace dict.iteritems() with dict.items()

"iteritems()" got omitted in python3.X

BonfaceKilz committed:
Remove relative imports

* wqflask/runserver.py: Relative imports leads to import errors.

BonfaceKilz committed:
Remove python2 string processing to utf-8

* wqflask/base/trait.py: Remove python2 codecs.BOM_UTF8 methods.

* wqflask/wqflask/search_results.py: Remove utf-8 decode method on string.

BonfaceKilz committed:
Update imports

* wqflask/wqflask/api/router.py: Break imports into separate lines and remove
unused ones.

BonfaceKilz committed:
Add a check for a "uuid" object

* wqflask/wqflask/views.py(json_default_handler): Check for "uuid.UUID" instance
to avoid TypeError.

BonfaceKilz committed:
Add check for empty group_code when adding a new resource

* wqflask/utility/authentication_tools.py(add_new_resource): If group_code is
"None", an error is thrown when you try to: `group_code + "_" + str(trait_id)`

BonfaceKilz committed:
Merge branch 'testing' into build/python3-migration

BonfaceKilz commented:
@zsloan could you have a look at that :point_up_2: ? The commits are pretty atomic so they should be easy to follow. Also @pjotrp and @robwwilliams  I'll create a separate issue to add [type hinting](https://docs.python.org/3/library/typing.html) and to use mypy to clean things up. We are on 3.8 so we may as well use it :smile:

robwwilliams commented:
Good work and good to have type hints

On Fri, Aug 28, 2020 at 7:02 AM BonfaceKilz <notifications@github.com>
wrote:

> @zsloan <https://github.com/zsloan> could you have a look at that 👆 ?
> The commits are pretty atomic so they should be easy to follow. Also
> @pjotrp <https://github.com/pjotrp> and @robwwilliams
> <https://github.com/robwwilliams> I'll create a separate issue to add type
> hinting <https://docs.python.org/3/library/typing.html> and to use mypy
> to clean things up. We are on 3.8 so we may as well use it 😄
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/genenetwork/genenetwork2/pull/422#issuecomment-682487395>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AC55V7EKS4M7H2GMS2ZXEA3SC6MGZANCNFSM4QDQRVVQ>
> .
>
-- 
Rob

Robert W. Williams, Ph.D.
Chair: Department of Genetics, Genomics and Informatics
University of Tennessee Health Science Center
71 S Manassas St, Memphis TN 38103-3310,  TSRB Room 405
Tele/Cell: 901 604 4752
EMAIL: rwilliams@uthsc.edu  or  labwilliams@gmail.com

zsloan commented:
I'm guessing some of these changes require updating to Python 3? I tried
checking out the pull request but ran into an import error in
wqflask/wqflask/__init__.py (when importing Flask), but that probably makes
sense if Python 3 is required.

On Fri, Aug 28, 2020 at 7:02 AM BonfaceKilz <notifications@github.com>
wrote:

> @zsloan <https://github.com/zsloan> could you have a look at that 👆 ?
> The commits are pretty atomic so they should be easy to follow. Also
> @pjotrp <https://github.com/pjotrp> and @robwwilliams
> <https://github.com/robwwilliams> I'll create a separate issue to add type
> hinting <https://docs.python.org/3/library/typing.html> and to use mypy
> to clean things up. We are on 3.8 so we may as well use it 😄
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/genenetwork/genenetwork2/pull/422#issuecomment-682487395>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AANQJGCJATGBNGA67IYAFNLSC6MGZANCNFSM4QDQRVVQ>
> .
>

pjotrp commented:
We are not on Python3 yet. It needs a new profile. @BonfaceKilz should I create a new profile that Zach can switch to? I mean, are we ready for this :)

zsloan commented:
Ah, I think the only thing is this one issue with the new qtlreaper where
the output files don't approximate Mb locations for pseudomarkers. I made a
change to Christian's code that technically fixes it, but I'm pretty sure
there's a better way to do it. Let me do a push request and ask Christian
if he knows of a better way to do the same thing I did.

This actually isn't really holding us back, though; it's a pretty minor
thing that I don't think anyone will notice (it doesn't cause an error - it
just makes the pseudomarkers not be displayed when doing Interval Mapping
with Mb positions). And it should be fixed soon once the changes I made, or
their improved version, are added to the new qtlreaper.

On Fri, Aug 28, 2020 at 1:24 PM Pjotr Prins <notifications@github.com>
wrote:

> We are not on Python3 yet. It needs a new profile. @BonfaceKilz
> <https://github.com/BonfaceKilz> should I create a new profile that Zach
> can switch to? I mean, are we ready for this :)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/genenetwork/genenetwork2/pull/422#issuecomment-683040526>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AANQJGDFKYJ4H3SSAOKZZ63SC7Y4ZANCNFSM4QDQRVVQ>
> .
>

BonfaceKilz commented:
Sorry for the late response. @pjotrp you should create a new profile for that. In that profile, use `python3-genenetwork2` package.

BonfaceKilz committed:
Merge branch 'testing' into build/python3-migration

pjotrp commented:
Makes sense ;). Zach, you can try to create a profile on tux01? Same instruction as before, but use python3-genenetwork2 instead of genenetwork2 for both the package (to install) and the new profile (so to keep the old profile for python2).

BonfaceKilz committed:
Remove typo at beginning of line

* wqflask/utility/authentication_tools.py: Remove accidental char at beginning
of line.

BonfaceKilz committed:
Merge branch 'testing' into build/python3-migration

BonfaceKilz commented:
@zsloan The variable `value` is not set here: https://github.com/genenetwork/genenetwork2/blob/8da6a70916d2cf18e476ab0adf47f802c481205d/wqflask/base/trait.py#L99 What should this be?

BonfaceKilz committed:
Apply pep8

* wqflask/base/trait.py: Apply pep8.
* wqflask/utility/authentication_tools.py: Ditto.

BonfaceKilz committed:
Fix errors in tests

* wqflask/tests/utility/test_authentication_tools.py
test_check_resource_availability_non_default_mask): Mock flask's
global 'g' variable properly.
* wqflask/tests/base/test_trait.py: Ditto.
* wqflask/tests/utility/test_authentication_tools.py: Ditto.

robwwilliams commented:
If we are talking about "cellid" this is may (?) be the identifier for
Affymetrix probe cells (a component part of a probe set). Used in a few
older but good data sets. But double-check me if kw.get('fullname'):
name2 = value.split("::")
if len(name2) == 2:
self.dataset, self.name = name2
# self.cellid is set to None above

On Thu, Sep 17, 2020 at 9:06 AM BonfaceKilz <notifications@github.com>
wrote:

> @zsloan <https://github.com/zsloan> The variable value is not set here:
> https://github.com/genenetwork/genenetwork2/blob/8da6a70916d2cf18e476ab0adf47f802c481205d/wqflask/base/trait.py#L99
> What should this be?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/genenetwork/genenetwork2/pull/422#issuecomment-694259974>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AC55V7GYOASPDXBY3IFM5D3SGIJXJANCNFSM4QDQRVVQ>
> .
>


-- 
Rob

Robert W. Williams, Ph.D.
Chair: Department of Genetics, Genomics and Informatics
University of Tennessee Health Science Center
71 S Manassas St, Memphis TN 38103-3310,  TSRB Room 405
Tele/Cell: 901 604 4752
EMAIL: rwilliams@uthsc.edu  or  labwilliams@gmail.com

BonfaceKilz committed:
Replace string arguments to "hmac.new" with bytearray

Same as:
https://github.com/genenetwork/genenetwork2/pull/422/commits/46443ec8d2cdfd7c60358a889d90a90e4f7daaf4

BonfaceKilz commented:
Currently getting the following tests failing (cc @zsloan  @pjotrp ):

<details>
<summary><b>Click to see output</b></summary>
<pre>

======================================================================
FAIL: test_data_hmac (tests.utility.test_hmac.TestHmacUtil)
Test data_hmac fn with a utf-8 string
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/bonface/projects/genenetwork2/wqflask/tests/utility/test_hmac.py", line 21, in test_data_hmac
    self.assertEqual(data_hmac("ファイ"), "ファイ:21fa1d935bbbb07a7875")
AssertionError: 'ファイ:b7b97e82167f93be8fc6' != 'ファイ:21fa1d935bbbb07a7875'
- ファイ:b7b97e82167f93be8fc6
+ ファイ:21fa1d935bbbb07a7875


======================================================================
FAIL: test_hmac_creation (tests.utility.test_hmac.TestHmacUtil)
Test hmac creation with a utf-8 string
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/bonface/projects/genenetwork2/wqflask/tests/utility/test_hmac.py", line 17, in test_hmac_creation
    self.assertEqual(hmac_creation("ファイ"), "21fa1d935bbbb07a7875")
AssertionError: 'b7b97e82167f93be8fc6' != '21fa1d935bbbb07a7875'
- b7b97e82167f93be8fc6
+ 21fa1d935bbbb07a7875


======================================================================
FAIL: test_url_for_hmac_with_param_in_url (tests.utility.test_hmac.TestHmacUtil)
Test url_for_hmac with params
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/bonface/.local/lib/python3.8/site-packages/mock/mock.py", line 1369, in patched
    return func(*newargs, **newkeywargs)
  File "/home/bonface/projects/genenetwork2/wqflask/tests/utility/test_hmac.py", line 34, in test_url_for_hmac_with_param_in_url                        
    self.assertEqual(url_for_hmac("ファイ"),
AssertionError: 'https://mock_url.com/?ファイ=1&hm=d38d74d8c8df998f9eff' != 'https://mock_url.com/?ファイ=1&hm=b2128fb28bc32da3b5b7'
- https://mock_url.com/?ファイ=1&hm=d38d74d8c8df998f9eff
+ https://mock_url.com/?ファイ=1&hm=b2128fb28bc32da3b5b7


======================================================================
FAIL: test_url_for_hmac_with_plain_url (tests.utility.test_hmac.TestHmacUtil)
Test url_for_hmac without params
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/bonface/.local/lib/python3.8/site-packages/mock/mock.py", line 1369, in patched
    return func(*newargs, **newkeywargs)
  File "/home/bonface/projects/genenetwork2/wqflask/tests/utility/test_hmac.py", line 27, in test_url_for_hmac_with_plain_url
    self.assertEqual(url_for_hmac("ファイ"),
AssertionError: 'https://mock_url.com/ファイ/?hm=a9d0d0a0f6a26c1e70fa' != 'https://mock_url.com/ファイ/?hm=a62896a50d9ffcff7deb'
- https://mock_url.com/ファイ/?hm=a9d0d0a0f6a26c1e70fa
+ https://mock_url.com/ファイ/?hm=a62896a50d9ffcff7deb


----------------------------------------------------------------------
Ran 72 tests in 0.066s

FAILED (failures=4)
</pre>
</details>

FYI, the above passes in the *testing* branch. Trying to figure out why the hash is returning something different. This is probably related to: https://github.com/genenetwork/genenetwork2/issues/440

BonfaceKilz commented:
robwwilliams <notifications@github.com> writes:

> If we are talking about "cellid" this is may (?) be the identifier for
> Affymetrix probe cells (a component part of a probe set). Used in a few
> older but good data sets. But double-check me if kw.get('fullname'):
> name2 = value.split("::")
> if len(name2) == 2:
> self.dataset, self.name = name2
> # self.cellid is set to None above
>

Yeah. Thanks for pointing that out. @zsloan is
best placed to clarify that. I don't think that if
branch has been hit(otherwise users would have
complained).

> On Thu, Sep 17, 2020 at 9:06 AM BonfaceKilz <notifications@github.com>
> wrote:
>
>> @zsloan <https://github.com/zsloan> The variable value is not set here:
>> https://github.com/genenetwork/genenetwork2/blob/8da6a70916d2cf18e476ab0adf47f802c481205d/wqflask/base/trait.py#L99
>> What should this be?
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/genenetwork/genenetwork2/pull/422#issuecomment-694259974>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AC55V7GYOASPDXBY3IFM5D3SGIJXJANCNFSM4QDQRVVQ>
>> .
>>
>
> --
> Rob
>
> Robert W. Williams, Ph.D.
> Chair: Department of Genetics, Genomics and Informatics
> University of Tennessee Health Science Center
> 71 S Manassas St, Memphis TN 38103-3310, TSRB Room 405
> Tele/Cell: 901 604 4752
> EMAIL: rwilliams@uthsc.edu or labwilliams@gmail.com
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.*

-- 
Bonface M. K. (https://www.bonfacemunyoki.com)
Chief Emacs Mchochezi
GPG key = D4F09EB110177E03C28E2FE1F5BBAE1E0392253F

BonfaceKilz committed:
Merge branch 'testing' into build/python3-migration

BonfaceKilz commented:
Fixed the failing tests by mocking `utility.hmac.app.config['SECRET_HMAC_CODE']` in the tests. See: https://github.com/genenetwork/genenetwork2/commit/0fd23995fa21c32c2a6005b2fd70b0bd99d13214. So that means how we were reading that value somehow changed(?) between Python2 and Python3.

zsloan commented:
I think the value thing might actually be some left-over logic from GN1
that is never hit because "fullname" is never passed as a parameter into
trait.py.

Just to make sure I did a grep of fullname across the code base, and it
doesn't seem like that's ever passed in when defining the trait. So it's
probably fine to remove it.

On Thu, Sep 17, 2020 at 10:12 AM BonfaceKilz <notifications@github.com>
wrote:

> Fixed the failing tests by mocking
> utility.hmac.app.config['SECRET_HMAC_CODE'] in the tests. See: 0fd2399
> <https://github.com/genenetwork/genenetwork2/commit/0fd23995fa21c32c2a6005b2fd70b0bd99d13214>.
> So that means how we were reading that value somehow changed(?) between
> Python2 and Python3.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/genenetwork/genenetwork2/pull/422#issuecomment-694303364>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AANQJGGESMVT4MR6PJQBUHDSGIRPRANCNFSM4QDQRVVQ>
> .
>

BonfaceKilz committed:
Convert "target_string" to a str type

* wqflask/base/trait.py (retrieve_trait_info): Convert "target_string"
to a str, otherwise an exception will be thrown if it's None.

BonfaceKilz committed:
Update Mechanical Rob

* test/requests/link_checker.py (verify_static_file): Cast
result.content.find's args to bytes.
* test/requests/main_web_functionality.py (check_search_page): Update
text to search for in a successful result. This value changed in
2aac6a06.

BonfaceKilz commented:
zsloan <notifications@github.com> writes:

> I think the value thing might actually be some left-over logic from GN1
> that is never hit because "fullname" is never passed as a parameter into
> trait.py.
>
> Just to make sure I did a grep of fullname across the code base, and it
> doesn't seem like that's ever passed in when defining the trait. So it's
> probably fine to remove it.
>

Thanks for the pointer. Could you remove it on the
CI branch?

> On Thu, Sep 17, 2020 at 10:12 AM BonfaceKilz <notifications@github.com>
> wrote:
>
>> Fixed the failing tests by mocking
>> utility.hmac.app.config['SECRET_HMAC_CODE'] in the tests. See: 0fd2399
>> <https://github.com/genenetwork/genenetwork2/commit/0fd23995fa21c32c2a6005b2fd70b0bd99d13214>.
>> So that means how we were reading that value somehow changed(?) between
>> Python2 and Python3.
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/genenetwork/genenetwork2/pull/422#issuecomment-694303364>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AANQJGGESMVT4MR6PJQBUHDSGIRPRANCNFSM4QDQRVVQ>
>> .
>>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.*

-- 
Bonface M. K. (https://www.bonfacemunyoki.com)
Chief Emacs Mchochezi
GPG key = D4F09EB110177E03C28E2FE1F5BBAE1E0392253F

BonfaceKilz committed:
Merge branch 'testing' into build/python3-migration

BonfaceKilz committed:
Fix inconsistent line ending in file

* wqflask/wqflask/db_info.py: Replace ^M with proper unix line
endings. See <http://ergoemacs.org/emacs/emacs_line_ending_char.html>

BonfaceKilz committed:
Run 2to3-8 tool through file

* wqflask/wqflask/db_info.py:
Run: 2to3-3.8 -w wqflask/wqflask/db_info.py

BonfaceKilz committed:
Apply PEP 8 to buffer

* wqflask/wqflask/db_info.py: Run pep8 through buffers. Fixes
inconsistent use of tabs and spaces in indentation which trips up
unittests.

BonfaceKilz committed:
Fix failing tests

* wqflask/tests/base/test_trait.py: Remove decode function call which
is no longer supported in python3.

BonfaceKilz committed:
Use Python3 image to run tests

BonfaceKilz committed:
Fix failing tests by replacing mock with unittest.mock

* wqflask/tests/base/test_data_set.py (imports): Run:
`find . -type f -name "*py" -print0 | xargs -0 sed -i \
"s|import mock|from unittest import mock|g"`
* wqflask/tests/base/test_trait.py(imports): Ditto.
* wqflask/tests/utility/test_authentication_tools.py(imports): Ditto.
* wqflask/tests/utility/test_hmac.py(imports). Ditto.
* wqflask/tests/wqflask/api/test_gen_menu.py: Ditto

BonfaceKilz committed:
Fix casting error

* wqflask/base/trait.py (retrieve_trait_info): If the
description_string or discription_display value return a None object,
evaluate to an empty string. Fixes errors related to *len(None)*.

BonfaceKilz committed:
Merge branch 'testing' into build/python3-migration

BonfaceKilz committed:
Merge branch 'build/python3-migration' of github.com:BonfaceKilz/genenetwork2 into build/python3-migration

BonfaceKilz committed:
Fix failing test

* wqflask/tests/base/test_trait.py: Add app_context to test class,
otherwise an error related to working outside the app_context is
generated. Adding the app_context is required when mocking "Flask.g".

BonfaceKilz commented:
@robwwilliams could you have a look at: http://gn2-test2.genenetwork.org. That runs the Python3 code. AFAICT this is ready for merging. Perhaps you'd catch something I never noticed :shrug: Other than that, I'll investigate: https://github.com/genenetwork/genenetwork2/issues/440. @pjotrp and @zsloan you could help too.

BonfaceKilz committed:
Add test for process_traits from collections.py

* wqflask/tests/wqflask/test_collect.py: Add it.

BonfaceKilz committed:
Add check for bytes

* wqflask/wqflask/collect.py (process_traits): Check whether the
passed "unprocessed_traits" is a byte otherwise an error is generated
when bytes(the default) are passed.

BonfaceKilz committed:
Apply pep 8 to buffer

* wqflask/wqflask/user_manager.py:
Remove some stale comments.
(imports): Apply pep-8 and remove unused imports.

BonfaceKilz committed:
Remove undefined variable

BonfaceKilz committed:
Fix import in tests

BonfaceKilz commented:
@zsloan in Python3, the global context `g` (even though it's bound to the application context) appears to be reset in each request, thereby having different `g.user_session.user_id` values between different requests .i.e. "/collections/new" and "/collections/views"; This leads to an error that looks like:

![image](https://user-images.githubusercontent.com/11820306/94629773-60781b00-02cc-11eb-8903-9d9070e4676b.png)

I want to try to work around it tomorrow, but if it's something you think would be easy for you to do/ solve, could you have a look at it from this branch? For reference, I've been looking at **collect.py** and **user_session.py**

PS: Having global state is almost always a bad idea.

zsloan commented:
Is it possible that Python 3 is returning StopIteration just due to hitting
the end of the tuple (I think that's a tuple) and Python 2.7 just didn't
return anything in that situation?

As for the global user_session, what would be a good way to make it not be
a global state? In this case I guess the main thing that would need to be
accessible almost everywhere is the user ID (I think that everything else
could then be fetched as needed using that). I'm not invested in it being
the way it is now, but I don't know how difficult it might be to change
(it's basically just like this because this is how it was designed way back
when the basic GN2 skeleton was created).



On Tue, Sep 29, 2020 at 7:39 PM BonfaceKilz <notifications@github.com>
wrote:

> @zsloan <https://github.com/zsloan> in Python3, the global context g
> (even though it's bound to the application context) appears to be reset in
> each request, thereby having different g.user_session.user_id values
> between different requests .i.e. "/collections/new" and
> "/collections/views"; This leads to an error that looks like:
>
> [image: image]
> <https://user-images.githubusercontent.com/11820306/94629773-60781b00-02cc-11eb-8903-9d9070e4676b.png>
>
> I want to try to work around it tomorrow, but if it's something you think
> would be easy for you to do/ solve, could you have a look at it from this
> branch? For reference, I've been looking at *collect.py* and
> *user_session.py*
>
> PS: Having global state is almost always a bad idea.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/genenetwork/genenetwork2/pull/422#issuecomment-701082458>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AANQJGFN7VQNVYFIL5CZKALSIJ44LANCNFSM4QDQRVVQ>
> .
>

BonfaceKilz commented:
zsloan <notifications@github.com> writes:

> Is it possible that Python 3 is returning StopIteration just due to hitting
> the end of the tuple (I think that's a tuple) and Python 2.7 just didn't
> return anything in that situation?

That's not really the problem. I did some
instrumentation on that front
yesterday. StopIteration is (rightly so) thrown
because we call *next* on an empty tuple--
*g.user_session.user_collections* returns an empty
tuple. *g.user_session.user_collections* fetches
the collections using the *user_id* which in this
branch changes on each request, thereby returning
an empty tuple.

> As for the global user_session, what would be a good way to make it not be
> a global state?

I'd have to give this some thought. Normally, when
figuring these things out, I normally try to use
"tests" as a "design tool". In this case, using
flask sessions who'd suffice(and they are easier
to test IMHO). But I'd keep that very
minimal. Sorry for the vague hand-wavy answer.

> In this case I guess the main thing that would need to be
> accessible almost everywhere is the user ID (I think that everything else
> could then be fetched as needed using that). I'm not invested in it being
> the way it is now, but I don't know how difficult it might be to change
> (it's basically just like this because this is how it was designed way back
> when the basic GN2 skeleton was created).
>

Oof! If the situation demands it, we'd have to
refactor it. But for now, I'm aiming to have
something that works without breaking things.

> On Tue, Sep 29, 2020 at 7:39 PM BonfaceKilz <notifications@github.com>
> wrote:
>
>> @zsloan <https://github.com/zsloan> in Python3, the global context g
>> (even though it's bound to the application context) appears to be reset in
>> each request, thereby having different g.user_session.user_id values
>> between different requests .i.e. "/collections/new" and
>> "/collections/views"; This leads to an error that looks like:
>>
>> [image: image]
>> <https://user-images.githubusercontent.com/11820306/94629773-60781b00-02cc-11eb-8903-9d9070e4676b.png>
>>
>> I want to try to work around it tomorrow, but if it's something you think
>> would be easy for you to do/ solve, could you have a look at it from this
>> branch? For reference, I've been looking at *collect.py* and
>> *user_session.py*
>>
>> PS: Having global state is almost always a bad idea.
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/genenetwork/genenetwork2/pull/422#issuecomment-701082458>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AANQJGFN7VQNVYFIL5CZKALSIJ44LANCNFSM4QDQRVVQ>
>> .
>>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.*

-- 
Bonface M. K. (https://www.bonfacemunyoki.com)
Chief Emacs Mchochezi / Twitter: @BonfaceKilz
GPG key = D4F09EB110177E03C28E2FE1F5BBAE1E0392253F

BonfaceKilz committed:
Fix StopIteration error thrown in view collections page

* wqflask/wqflask/user_session.py (user_id): Use byte-string when
fetching user_id. See:
https://github.com/genenetwork/genenetwork2/pull/422#issuecomment-701082458

BonfaceKilz commented:
@zsloan turns out I was looking at things the wrong way. Turns out the "user_id" variable is a byte! So I just did a simple fix(see: https://github.com/genenetwork/genenetwork2/pull/422/commits/dce0cd8725e68e6984c8b54a2cc108e4f8c9df7a) which works! No need to rethink how we fetch the user_id! :tada: :tada:

BonfaceKilz committed:
Apply pep 8 formatting

* wqflask/wqflask/correlation/show_corr_results.py
[imports]: Apply pep8 and remove unused imports.

BonfaceKilz committed:
Add missing import

* wqflask/wqflask/correlation/show_corr_results.py: Add it.

BonfaceKilz committed:
Make object JSON serializable

* wqflask/wqflask/search_results.py
(SearchResultPage.gen_search_result): Some of the items is
self.trait_list are in bytes, and *json.loads* can't serialize bytes
into JSON.

BonfaceKilz committed:
Convert trait attributes to utf-8 strings if they were bytes

BonfaceKilz committed:
Cast "previous_chr" to an int

* wqflask/wqflask/marker_regression/gemma_mapping.py
(parse_loco_output): Cast "previous_chr" to an int which prevents the
error "Python typeerror: ‘>’ not supported between instances of ‘str’
and ‘int’" from being thrown. The error is **only** thrown in Python3.

BonfaceKilz committed:
Remove unnecessary commented out blocks

* wqflask/wqflask/show_trait/SampleList.py: See above.

BonfaceKilz committed:
Update natural_sort function

* wqflask/wqflask/show_trait/SampleList.py
(natural_sort): Add new function.
(SampleList.get_attributes): Use new sort function to do natural_sort.

BonfaceKilz committed:
Replace is_integer with isinstance when checking for an int value

BonfaceKilz committed:
Cast variable properly to avoid type exceptions

* wqflask/wqflask/marker_regression/run_mapping.py: See 94c36c8c.

BonfaceKilz committed:
Apply PEP 8 formatting

BonfaceKilz committed:
Merge branch 'testing' of github.com:genenetwork/genenetwork2 into build/python3-migration

BonfaceKilz committed:
Split author's list into it's own variable

* wqflask/wqflask/templates/correlation_page.html: Ditto. Prevents
jinja from throwing an error due to compaing a list type with an int.

BonfaceKilz committed:
Ignore utf-8 encoding errors

* wqflask/base/trait.py (retrieve_trait_info): Add extra param to
ignore errors such as: "utf-8 codec can't decode byte 0xb5 in position
0: invalid start byte" thrown in Python3.

BonfaceKilz committed:
Apply PEP-8 formatting

BonfaceKilz committed:
Convert cmp function into a key function

* wqflask/wqflask/correlation_matrix/show_corr_matrix.py
(sortEigenVectors): Convert the comparison function into a key
function. See:
https://docs.python.org/3/howto/sorting.html#the-old-way-using-the-cmp-parameter

BonfaceKilz committed:
Update base64 conversion

* wqflask/wqflask/wgcna/wgcna_analysis.py
(WGCNA.render_image): Use Python3's base64 library from the stdlib to
perform conversion.
* wqflask/wqflask/ctl/ctl_analysis.py (CTL.loadImage): Ditto.

BonfaceKilz committed:
Use built-in array as a named import

* wqflask/wqflask/wgcna/wgcna_analysis.py
(WGCNA.render_image): Use named import method. This is a workaround
for using explicitly using "array" which conflicts with something else
in the namespace.

BonfaceKilz committed:
Replace "ncores" with "nthreads" param to r_CTLscan

BonfaceKilz committed:
Replace string concat("+") with python3's f-string

BonfaceKilz committed:
Update base64 conversion

BonfaceKilz committed:
Fix export function on traits and analysis page

* wqflask/wqflask/views.py (export_trait_excel): Write data as bytes
instead of strings. See:
https://stackoverflow.com/questions/32075135/python-3-in-memory-zipfile-error-string-argument-expected-got-bytes
* wqflask/wqflask/show_trait/export_trait_data.py: Convert cmp
function into a key function.

BonfaceKilz committed:
Merge branch 'testing' into build/python3-migration

BonfaceKilz committed:
Wrap args of print statements correctly using braces

* wqflask/wqflask/wgcna/wgcna_analysis.py: Update print statements by
wrapping then in "()". Run:
"2to3-3.8 -w wqflask/wqflask/wgcna/wgcna_analysis.py"

BonfaceKilz committed:
Delete stale comment

BonfaceKilz committed:
Update snpString link

* wqflask/wqflask/marker_regression/display_mapping_results.py:
Replace "%" strings with f-strings. Also, replace old "Href" with new
htmlgen "Link" method.

BonfaceKilz committed:
Add pbkdf2 tests

* wqflask/tests/wqflask/test_pbkdf2.py: New tests.

BonfaceKilz committed:
Update pbkdf2

* wqflask/wqflask/pbkdf2.py: Remove unnecessary pbkdf2 usage docs---
it already exists in python3 stdlib--- and "__main__" section which
runs the test.
(imports): Delete unused imports. Add safe_str_cmp.
(pbkdf2_hex): Use hashlib.pbkdf2_hmac.
(safe_str_cmp): Replace with werkzeug.security.safe_str_cmp
(test): Delete it. Tests moved to test_pbkdf2.py.

BonfaceKilz committed:
Add tests for encoding password

* wqflask/tests/wqflask/test_user_login.py: New tests.

BonfaceKilz committed:
Update how pbkdf2.pbkdf2_hex is used when encrypting pass

* wqflask/wqflask/user_login.py (encode_password):
- Remove unused local variable "salt".
- Pass "hashfunc" as a string.
- Apply Pep-8.

BonfaceKilz committed:
Merge branch 'testing' into build/python3-migration

zsloan commented:
It's working well for me now; I haven't noticed any other issues and I think I've tried most features.

I think it's probably good enough that, even if something does pop up, it won't be anything significant. We might still want Rob to use my branch some first while it's switched to using the Python 3 version, though.

zsloan reviewed (approved):
Oh, I just realized that I specifically comment/approve as part of reviewing changes.

BonfaceKilz commented:
zsloan <notifications@github.com> writes:

> @zsloan approved this pull request.
>
> Oh, I just realized that I specifically comment/approve as part of reviewing changes.
Yeah. But normally when a PR is on-going, I prefer
commenting or requesting changes as soon as I see
extra commits--- that way, I only get to review a
subset of changes instead of reviewing a whole set
of changes in one full swoop(it can be daunting at
times). Note that GH has the "Show changes since
your last review" under the "Files Changed" tab. I
make heavy use of that :)

-- 
Bonface M. K. <https://www.bonfacemunyoki.com>
Chief Emacs Bazu / Rieng ya software sare
Mchochezi of: <https://upbookclub.com> / Twitter: @BonfaceKilz
GPG Key: D4F09EB110177E03C28E2FE1F5BBAE1E0392253F

zsloan commented:
Yeah, that definitely makes more sense for PRs with a lot of changes like
this.

On Tue, Oct 20, 2020 at 5:17 PM BonfaceKilz <notifications@github.com>
wrote:

> zsloan <notifications@github.com> writes:
>
> > @zsloan approved this pull request.
> >
> > Oh, I just realized that I specifically comment/approve as part of
> reviewing changes.
> Yeah. But normally when a PR is on-going, I prefer
> commenting or requesting changes as soon as I see
> extra commits--- that way, I only get to review a
> subset of changes instead of reviewing a whole set
> of changes in one full swoop(it can be daunting at
> times). Note that GH has the "Show changes since
> your last review" under the "Files Changed" tab. I
> make heavy use of that :)
>
> --
> Bonface M. K. <https://www.bonfacemunyoki.com>
> Chief Emacs Bazu / Rieng ya software sare
> Mchochezi of: <https://upbookclub.com> / Twitter: @BonfaceKilz
> GPG Key: D4F09EB110177E03C28E2FE1F5BBAE1E0392253F
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/genenetwork/genenetwork2/pull/422#issuecomment-713172068>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AANQJGBASVQEIOWBR4YTU5TSLYD7XANCNFSM4QDQRVVQ>
> .
>

pjotrp commented:
Great! Let's switch over formally and move it to production in the next weeks. I agree that we can afford minor mishaps :)

robwwilliams commented:
Me three. Yes, minor mishaps like Microsoft, Facebook, Twitter.... ;-)

On Wed, Oct 21, 2020 at 3:17 AM Pjotr Prins <notifications@github.com>
wrote:

> Great! Let's switch over formally and move it to production in the next
> weeks. I agree that we can afford minor mishaps :)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/genenetwork/genenetwork2/pull/422#issuecomment-713394985>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AC55V7AK35LLQBELNFOIJFTSL2KITANCNFSM4QDQRVVQ>
> .
>


-- 
Rob

Robert W. Williams, Ph.D.
Chair: Department of Genetics, Genomics and Informatics
University of Tennessee Health Science Center
71 S Manassas St, Memphis TN 38103-3310,  TSRB Room 405
Tele/Cell: 901 604 4752
EMAIL: rwilliams@uthsc.edu  or  labwilliams@gmail.com

BonfaceKilz commented:
@pjotrp I think we should go ahead and merge this. I haven't seen any complaints so far. That plus on the python3 branch, all the tests pass without the weird test failures from the python2 testing branch. @robwwilliams, @zsloan and @pjotrp wdyt?

pjotrp commented:
Yeah, let's switch over :+1:

BonfaceKilz commented:
Cool. Optimistically merging this. Will revert if we run into problems we can't fix :smiley:
